### PR TITLE
feat: Support multiple regionwise prompts

### PR DESCRIPTION
**Pull request by rvaibhav**
Command to run

```
PYTHONPATH=$(pwd) CHECKPOINT_DIR="/config/models/cosmos-transfer1" python cosmos_transfer1/diffusion/inference/transfer.py \
    --checkpoint_dir /config/models/cosmos-transfer1 \
    --video_save_folder outputs/example1_single_control_segment_regional_prompt_three \
    --controlnet_specs assets/inference_cosmos_transfer1_single_control_segment_regional_prompt_three.json \
    --offload_text_encoder_model --sigma_max 80 --offload_guardrail_models
```

### CODE REVIEW

**Comment by rvaibhav** [comment]
Segment masks generated for configs `assets/inference_cosmos_transfer1_single_control_segment_regional_prompt_three.json`
[Bug]

**Left woman mask**

https://github.com/user-attachments/assets/f3d94101-cfb9-4b2f-b574-d6e20438bd5a


**Right Woman Mask**


https://github.com/user-attachments/assets/8f0b5056-b6fc-4d2c-9d48-940ec808bbf2

File: cosmos_transfer1/diffusion/inference/world_generation_pipeline.py [hunk 1]
```diff
@@ -635,7 +639,7 @@ def generate(
         log.info("Finish generation")
 
         log.info("Run guardrail on generated video")
-        video = self._run_guardrail_on_video_with_offload(video)
+        # video = self._run_guardrail_on_video_with_offload(video)
```
Comment (rvaibhav) [inline cosmos_transfer1/diffusion/inference/world_generation_pipeline.py]: Re-enable guardrails

File: cosmos_transfer1/diffusion/module/attention_i4.py [hunk 2]
```diff
@@ -0,0 +1,820 @@
+# -----------------------------------------------------------------------------
```
Comment (rvaibhav) [inline cosmos_transfer1/diffusion/module/attention_i4.py]: This file needs to be deleted.

File: cosmos_transfer1/auxiliary/sam2/sam2_model.py [hunk 3]
```diff
@@ -363,6 +363,7 @@ def __call__(
                     if binarize_video:
                         frames = np.any(frames > 0, axis=-1).astype(np.uint8) * 255
                     write_video(frames, output_video, fps)
+
```
Comment (rvaibhav) [inline cosmos_transfer1/auxiliary/sam2/sam2_model.py]: revert

File: cosmos_transfer1/utils/regional_prompting_utils.py [hunk 4]
```diff
@@ -157,15 +172,71 @@ def prepare_regional_prompts(
...
+        elif isinstance(region_definition, list):
+            region_definitions_list.append(region_definition)
+            box_prompts.append(regional_prompts[prompt_idx])
+        else:
+            raise ValueError(f"Region definition format not recognized: {type(region_definition)}")
+        prompt_idx += 1
+
+    # Update regional_prompts to maintain correct ordering
+    regional_prompts = box_prompts + seg_prompts
+    # segmentation_maps: List[torch.Tensor] = []
```
Comment (rvaibhav) [inline cosmos_transfer1/utils/regional_prompting_utils.py]: remove

File: cosmos_transfer1/utils/regional_prompting_utils.py [hunk 5]
```diff
@@ -157,15 +172,71 @@ def prepare_regional_prompts(
...
+    #     else:
+    #         raise ValueError(f"Region definition format not recognized: {type(region_definition)}")
+    region_masks_boxes = processor.create_region_masks_from_boxes(
+        region_definitions_list, batch_size, time_dim, height, width, device
+    )
+    region_masks_segmentation = processor.create_region_masks_from_segmentation(
+        segmentation_maps, batch_size, time_dim, height, width, device
+    )
+    region_masks = torch.cat([region_masks_boxes, region_masks_segmentation], dim=1)
+    # if isinstance(region_definitions[0], list):
```
Comment (rvaibhav) [inline cosmos_transfer1/utils/regional_prompting_utils.py]: remove commented code

File: cosmos_transfer1/diffusion/inference/preprocessors.py [hunk 6]
```diff
@@ -28,10 +30,11 @@ def __init__(self):
         self.seg_model = None
         self.keypoint_model = None
 
-    def __call__(self, input_video, input_prompt, control_inputs, output_folder):
+    def __call__(self, input_video, input_prompt, control_inputs, output_folder, regional_prompts=None):
+
         for hint_key in control_inputs:
             if hint_key in ["depth", "seg", "keypoint"]:
-                self.gen_input_control(input_video, input_prompt, hint_key, control_inputs[hint_key], output_folder)
+                self.gen_input_control(input_video, input_prompt, hint_key, control_inputs[hint_key], output_folder, regional_prompts)
```
Comment (rvaibhav) [inline cosmos_transfer1/diffusion/inference/preprocessors.py]: regional_prompts not needed for `gen_input_control`

File: cosmos_transfer1/diffusion/inference/preprocessors.py [hunk 7]
```diff
@@ -57,9 +60,27 @@ def __call__(self, input_video, input_prompt, control_inputs, output_folder):
...
+                            prompt=prompt,
+                            weight_scaler=1.0,
+                            binarize_video=True,
+                        )
+                        if os.path.exists(out_tensor):
+                            regional_prompt["region_definitions_path"] = out_tensor
         return control_inputs
 
-    def gen_input_control(self, in_video, in_prompt, hint_key, control_input, output_folder):
+    def gen_input_control(self, in_video, in_prompt, hint_key, control_input, output_folder, regional_prompts=None):
```
Comment (rvaibhav) [inline cosmos_transfer1/diffusion/inference/preprocessors.py]: revert definition change for `gen_input_control`

File: cosmos_transfer1/diffusion/inference/preprocessors.py [hunk 8]
```diff
@@ -57,9 +60,27 @@ def __call__(self, input_video, input_prompt, control_inputs, output_folder):
...
+                        prompt = regional_prompt["mask_prompt"]
+                        out_tensor = os.path.join(output_folder, f"regional_context_{i}.pt")
+                        out_video = os.path.join(output_folder, f"regional_context_{i}.mp4")
+                        self.segmentation(
+                            in_video=input_video,
+                            out_tensor=out_tensor,
+                            out_video=out_video,
+                            prompt=prompt,
+                            weight_scaler=1.0,
+                            binarize_video=True,
```
Comment (pjannaty) [inline cosmos_transfer1/diffusion/inference/preprocessors.py]: This does not achieve what we hope achieve. Looking at the function, it first runs the usual multi-color segmentation, then tries to convert the overall multi-color image to a binary image. In the multi-color segmentation, both women are highlighted. Who becomes dark and who becomes white depends on the shade of the color.

We should instead use the first selection `boxes[0]` as-in [here](https://gitlab-master.nvidia.com/dir/nvidia-cosmos/cosmos-transfer1/-/merge_requests/149#note_34999541).

Comment (rvaibhav) [reply cosmos_transfer1/diffusion/inference/preprocessors.py]: Addressed
